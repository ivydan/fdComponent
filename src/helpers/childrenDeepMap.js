import { Children, cloneElement } from 'react';
import Tab from '../components/Tab';
import TabList from '../components/TabList';
import TabPanel from '../components/TabPanel';

function isTabChild(child) {
  return child.type === Tab || child.type === TabList || child.type === TabPanel;
}

export function deepMap(children, callback) {
  return Children.map(children, child => {
    // null happens when conditionally rendering TabPanel/Tab
    // see https://github.com/reactjs/react-tabs/issues/37
    if (child === null) return null;

    if (isTabChild(child)) {
      return callback(child);
    }

    if (child.props && child.props.children && typeof child.props.children === 'object') {
      // Clone the child that has children and map them too
      return cloneElement(child, {
        ...child.props,
        children: deepMap(child.props.children, callback),
      });
    }

    return child;
  });
}

export function deepForEach(children, callback) {
  return Children.forEach(children, child => {
    // null happens when conditionally rendering TabPanel/Tab
    // see https://github.com/reactjs/react-tabs/issues/37
    if (child === null) return;

    if (child.type === Tab || child.type === TabPanel) {
      callback(child);
    } else if (child.props && child.props.children && typeof child.props.children === 'object') {
      if (child.type === TabList) callback(child);
      deepForEach(child.props.children, callback);
    }
  });
}
