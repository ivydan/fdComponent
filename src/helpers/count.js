import { deepForEach } from '../helpers/childrenDeepMap';
import Tab from '../components/Tab';
import TabPanel from '../components/TabPanel';

export function getTabsCount(children) {
  let tabCount = 0;
  deepForEach(children, child => {
    if (child.type === Tab) tabCount++;
  });

  return tabCount;
}

export function getPanelsCount(children) {
  let panelCount = 0;
  deepForEach(children, child => {
    if (child.type === TabPanel) panelCount++;
  });

  return panelCount;
}
