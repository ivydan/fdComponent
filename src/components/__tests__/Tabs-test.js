/* eslint-env jest */
/* eslint-disable react/no-multi-comp */
import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Tab from '../Tab';
import TabList from '../TabList';
import TabPanel from '../TabPanel';
import Tabs from '../Tabs';
import { reset as resetIdCounter } from '../../helpers/uuid';

function expectToMatchSnapshot(component) {
  expect(renderer.create(component).toJSON()).toMatchSnapshot();
}

function createTabs(props = {}) {
  return (
    <Tabs {...props}>
      <TabList>
        <Tab>Foo</Tab>
        <Tab>Bar</Tab>
        <Tab>
          <a>Baz</a>
        </Tab>
        <Tab disabled>Qux</Tab>
      </TabList>
      <TabPanel>Hello Foo</TabPanel>
      <TabPanel>Hello Bar</TabPanel>
      <TabPanel>Hello Baz</TabPanel>
      <TabPanel>Hello Qux</TabPanel>
    </Tabs>
  );
}

function assertTabSelected(wrapper, index) {
  const tab = wrapper.childAt(0).childAt(index);
  const panel = wrapper.childAt(index + 1);

  expect(tab.prop('selected')).toBe(true);
  expect(panel.prop('selected')).toBe(true);
}

describe('<Tabs />', () => {
  beforeEach(() => resetIdCounter());

  beforeAll(() => {
    // eslint-disable-next-line no-console
    console.error = error => {
      throw new Error(error);
    };
  });

  describe('props', () => {
    test('should have sane defaults', () => {
      expectToMatchSnapshot(createTabs());
    });

    test('should honor positive defaultIndex prop', () => {
      expectToMatchSnapshot(createTabs({ defaultIndex: 1 }));
    });

    test('should honor negative defaultIndex prop', () => {
      expectToMatchSnapshot(createTabs({ defaultIndex: -1 }));
    });

    test('should call onSelect when selection changes', () => {
      const called = { index: -1, last: -1 };
      const wrapper = mount(
        createTabs({
          onSelect(index, last) {
            called.index = index;
            called.last = last;
          },
        }),
      );

      wrapper.childAt(0).childAt(1).simulate('click');

      expect(called.index).toBe(1);
      expect(called.last).toBe(0);
    });

    test('should accept className', () => {
      expectToMatchSnapshot(createTabs({ className: 'foobar' }));
    });
  });

  describe('child props', () => {
    test('should reset ids correctly', () => {
      expectToMatchSnapshot(createTabs());

      resetIdCounter();

      expectToMatchSnapshot(createTabs());
    });
  });

  describe('interaction', () => {
    test('should update selectedIndex when clicked', () => {
      const wrapper = mount(createTabs());
      wrapper.childAt(0).childAt(1).simulate('click');

      assertTabSelected(wrapper, 1);
    });

    test('should update selectedIndex when tab child is clicked', () => {
      const wrapper = mount(createTabs());
      const tablist = wrapper.childAt(0);
      tablist.childAt(2).first().simulate('click');

      assertTabSelected(wrapper, 2);
    });

    test('should not change selectedIndex when clicking a disabled tab', () => {
      const wrapper = mount(createTabs({ defaultIndex: 0 }));

      wrapper.childAt(0).childAt(3).simulate('click');
      assertTabSelected(wrapper, 0);
    });
  });

  describe('performance', () => {
    test('should only render the selected tab panel', () => {
      const wrapper = mount(createTabs());

      expect(wrapper.childAt(1).text()).toBe('Hello Foo');
      expect(wrapper.childAt(2).text()).toBe('');
      expect(wrapper.childAt(3).text()).toBe('');

      wrapper.childAt(0).childAt(1).simulate('click');

      expect(wrapper.childAt(1).text()).toBe('');
      expect(wrapper.childAt(2).text()).toBe('Hello Bar');
      expect(wrapper.childAt(3).text()).toBe('');

      wrapper.childAt(0).childAt(2).simulate('click');

      expect(wrapper.childAt(1).text()).toBe('');
      expect(wrapper.childAt(2).text()).toBe('');
      expect(wrapper.childAt(3).text()).toBe('Hello Baz');
    });

    test('should render all tabs if forceRenderTabPanel is true', () => {
      expectToMatchSnapshot(createTabs({ forceRenderTabPanel: true }));
    });
  });

  describe('validation', () => {
    test('should result with warning when tabs/panels are imbalanced', () => {
      const oldConsoleError = console.error; // eslint-disable-line no-console
      console.error = () => {}; // eslint-disable-line no-console
      const wrapper = shallow(
        <Tabs>
          <TabList>
            <Tab>Foo</Tab>
          </TabList>
        </Tabs>,
      );
      console.error = oldConsoleError; // eslint-disable-line no-console

      const result = Tabs.propTypes.children(wrapper.props(), 'children', 'Tabs');
      expect(result instanceof Error).toBe(true);
    });

    test('should result with warning when tab outside of tablist', () => {
      const oldConsoleError = console.error; // eslint-disable-line no-console
      console.error = () => {}; // eslint-disable-line no-console
      const wrapper = shallow(
        <Tabs>
          <TabList>
            <Tab>Foo</Tab>
          </TabList>
          <Tab>Foo</Tab>
          <TabPanel />
          <TabPanel />
        </Tabs>,
      );
      console.error = oldConsoleError; // eslint-disable-line no-console

      const result = Tabs.propTypes.children(wrapper.props(), 'children', 'Tabs');
      expect(result instanceof Error).toBe(true);
    });

    test('should result with warning when multiple tablist components exist', () => {
      const oldConsoleError = console.error; // eslint-disable-line no-console
      console.error = () => {}; // eslint-disable-line no-console
      const wrapper = shallow(
        <Tabs>
          <TabList>
            <Tab>Foo</Tab>
          </TabList>
          <TabList>
            <Tab>Foo</Tab>
          </TabList>
          <TabPanel />
          <TabPanel />
        </Tabs>,
      );
      console.error = oldConsoleError; // eslint-disable-line no-console

      const result = Tabs.propTypes.children(wrapper.props(), 'children', 'Tabs');
      expect(result instanceof Error).toBe(true);
    });

    test('should result with warning when onSelect missing when selectedIndex set', () => {
      const oldConsoleError = console.error; // eslint-disable-line no-console
      const catchedErrors = [];
      // eslint-disable-next-line no-console
      console.error = error => {
        catchedErrors.push(error);
      };
      shallow(
        <Tabs selectedIndex={1}>
          <TabList>
            <Tab>Foo</Tab>
          </TabList>
          <TabPanel>Foo</TabPanel>
        </Tabs>,
      );
      console.error = oldConsoleError; // eslint-disable-line no-console

      const expectedMessage =
        'The prop `onSelect` is marked as required in `Tabs`, but its value is `undefined` or `null`.';
      expect(catchedErrors.some(msg => msg.indexOf(expectedMessage) > -1)).toBe(true);
    });

    test('should result with warning when defaultIndex and selectedIndex set', () => {
      const oldConsoleError = console.error; // eslint-disable-line no-console
      const catchedErrors = [];
      // eslint-disable-next-line no-console
      console.error = error => {
        catchedErrors.push(error);
      };
      shallow(
        <Tabs selectedIndex={1} defaultIndex={1}>
          <TabList>
            <Tab>Foo</Tab>
          </TabList>
          <TabPanel>Foo</TabPanel>
        </Tabs>,
      );
      console.error = oldConsoleError; // eslint-disable-line no-console

      const expectedMessage =
        'The prop `selectedIndex` cannot be used together with `defaultIndex` in `Tabs`.';
      expect(catchedErrors.some(msg => msg.indexOf(expectedMessage) > -1)).toBe(true);
    });

    test(`should result with warning when tabs/panels are imbalanced and
        it should ignore non tab children`, () => {
      const oldConsoleError = console.error; // eslint-disable-line no-console
      console.error = () => {}; // eslint-disable-line no-console
      const wrapper = shallow(
        <Tabs>
          <TabList>
            <Tab>Foo</Tab>
            <div>+</div>
          </TabList>

          <TabPanel>Hello Foo</TabPanel>
          <TabPanel>Hello Bar</TabPanel>
        </Tabs>,
      );
      console.error = oldConsoleError; // eslint-disable-line no-console

      const result = Tabs.propTypes.children(wrapper.props(), 'children', 'Tabs');
      expect(result instanceof Error).toBe(true);
    });

    test('should allow random order for elements', () => {
      expectToMatchSnapshot(
        <Tabs forceRenderTabPanel>
          <TabPanel>Hello Foo</TabPanel>
          <TabList>
            <Tab>Foo</Tab>
            <Tab>Bar</Tab>
          </TabList>
          <TabPanel>Hello Bar</TabPanel>
        </Tabs>,
      );
    });

    test('should not throw a warning when wrong element is found', () => {
      const wrapper = shallow(
        <Tabs>
          <TabList>
            <Tab />
            <div />
          </TabList>
          <TabPanel />
        </Tabs>,
      );

      const result = Tabs.propTypes.children(wrapper.props(), 'children', 'Tabs');
      expect(result instanceof Error).toBe(false);
    });

    test('should be okay with rendering without any children', () => {
      expect(() => shallow(<Tabs />)).not.toThrow();
    });

    test('should be okay with rendering just TabList', () => {
      expect(() =>
        shallow(
          <Tabs>
            <TabList />
          </Tabs>,
        ),
      ).not.toThrow();
    });

    test('should gracefully render null', () => {
      expect(() =>
        shallow(
          <Tabs>
            <TabList>
              <Tab>Tab A</Tab>
              {false && <Tab>Tab B</Tab>}
            </TabList>
            <TabPanel>Content A</TabPanel>
            {false && <TabPanel>Content B</TabPanel>}
          </Tabs>,
        ),
      ).not.toThrow();
    });

    test('should support nested tabs', () => {
      const wrapper = mount(
        <Tabs className="first">
          <TabList>
            <Tab />
            <Tab />
          </TabList>
          <TabPanel>
            <Tabs className="second">
              <TabList>
                <Tab />
                <Tab />
              </TabList>
              <TabPanel />
              <TabPanel />
            </Tabs>
          </TabPanel>
          <TabPanel />
        </Tabs>,
      );

      const innerTabs = wrapper.childAt(1).childAt(0);

      innerTabs.childAt(0).childAt(1).simulate('click');

      assertTabSelected(wrapper, 0);
      assertTabSelected(innerTabs, 1);
    });

    test('should allow other DOM nodes', () => {
      expectToMatchSnapshot(
        <Tabs>
          <div id="tabs-nav-wrapper">
            <button>Left</button>
            <div className="tabs-container">
              <TabList>
                <Tab />
                <Tab />
              </TabList>
            </div>
            <button>Right</button>
          </div>
          <div className="tab-panels">
            <TabPanel />
            <TabPanel />
          </div>
        </Tabs>,
      );
    });
  });

  test('should pass through custom properties', () => {
    expectToMatchSnapshot(<Tabs data-tooltip="Tooltip contents" />);
  });

  test('should not add known props to dom', () => {
    expectToMatchSnapshot(<Tabs defaultIndex={3} />);
  });

  test('should cancel if event handler returns false', () => {
    const wrapper = mount(createTabs({ onSelect: () => false }));

    assertTabSelected(wrapper, 0);

    wrapper.childAt(0).childAt(1).simulate('click');
    assertTabSelected(wrapper, 0);

    wrapper.childAt(0).childAt(2).simulate('click');
    assertTabSelected(wrapper, 0);
  });

  test('should switch tabs if setState is called within onSelect', () => {
    class Wrap extends React.Component {
      handleSelect = () => this.setState({ foo: 'bar' });
      render() {
        return createTabs({ onSelect: this.handleSelect });
      }
    }

    const wrapper = mount(<Wrap />);

    wrapper.childAt(0).childAt(1).simulate('click');
    assertTabSelected(wrapper, 1);

    wrapper.childAt(0).childAt(2).simulate('click');
    assertTabSelected(wrapper, 2);
  });
});
