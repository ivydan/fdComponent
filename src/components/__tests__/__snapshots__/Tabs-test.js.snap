// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> child props should reset ids correctly 1`] = `
<div
  className="react-tabs"
  data-tabs={true}
  onClick={[Function]}
  onKeyDown={[Function]}
>
  <ul
    className="react-tabs__tab-list"
    role="tablist"
  >
    <li
      aria-controls="react-tabs-1"
      aria-disabled="false"
      aria-selected="true"
      className="react-tabs__tab react-tabs__tab--selected"
      id="react-tabs-0"
      role="tab"
      tabIndex="0"
    >
      Foo
    </li>
    <li
      aria-controls="react-tabs-3"
      aria-disabled="false"
      aria-selected="false"
      className="react-tabs__tab"
      id="react-tabs-2"
      role="tab"
      tabIndex={null}
    >
      Bar
    </li>
    <li
      aria-controls="react-tabs-5"
      aria-disabled="false"
      aria-selected="false"
      className="react-tabs__tab"
      id="react-tabs-4"
      role="tab"
      tabIndex={null}
    >
      <a>
        Baz
      </a>
    </li>
    <li
      aria-controls="react-tabs-7"
      aria-disabled="true"
      aria-selected="false"
      className="react-tabs__tab react-tabs__tab--disabled"
      id="react-tabs-6"
      role="tab"
      tabIndex={null}
    >
      Qux
    </li>
  </ul>
  <div
    aria-labelledby="react-tabs-0"
    className="react-tabs__tab-panel react-tabs__tab-panel--selected"
    id="react-tabs-1"
    role="tabpanel"
    style={Object {}}
  >
    Hello Foo
  </div>
  <div
    aria-labelledby="react-tabs-2"
    className="react-tabs__tab-panel"
    id="react-tabs-3"
    role="tabpanel"
    style={Object {}}
  />
  <div
    aria-labelledby="react-tabs-4"
    className="react-tabs__tab-panel"
    id="react-tabs-5"
    role="tabpanel"
    style={Object {}}
  />
  <div
    aria-labelledby="react-tabs-6"
    className="react-tabs__tab-panel"
    id="react-tabs-7"
    role="tabpanel"
    style={Object {}}
  />
</div>
`;

exports[`<Tabs /> child props should reset ids correctly 2`] = `
<div
  className="react-tabs"
  data-tabs={true}
  onClick={[Function]}
  onKeyDown={[Function]}
>
  <ul
    className="react-tabs__tab-list"
    role="tablist"
  >
    <li
      aria-controls="react-tabs-1"
      aria-disabled="false"
      aria-selected="true"
      className="react-tabs__tab react-tabs__tab--selected"
      id="react-tabs-0"
      role="tab"
      tabIndex="0"
    >
      Foo
    </li>
    <li
      aria-controls="react-tabs-3"
      aria-disabled="false"
      aria-selected="false"
      className="react-tabs__tab"
      id="react-tabs-2"
      role="tab"
      tabIndex={null}
    >
      Bar
    </li>
    <li
      aria-controls="react-tabs-5"
      aria-disabled="false"
      aria-selected="false"
      className="react-tabs__tab"
      id="react-tabs-4"
      role="tab"
      tabIndex={null}
    >
      <a>
        Baz
      </a>
    </li>
    <li
      aria-controls="react-tabs-7"
      aria-disabled="true"
      aria-selected="false"
      className="react-tabs__tab react-tabs__tab--disabled"
      id="react-tabs-6"
      role="tab"
      tabIndex={null}
    >
      Qux
    </li>
  </ul>
  <div
    aria-labelledby="react-tabs-0"
    className="react-tabs__tab-panel react-tabs__tab-panel--selected"
    id="react-tabs-1"
    role="tabpanel"
    style={Object {}}
  >
    Hello Foo
  </div>
  <div
    aria-labelledby="react-tabs-2"
    className="react-tabs__tab-panel"
    id="react-tabs-3"
    role="tabpanel"
    style={Object {}}
  />
  <div
    aria-labelledby="react-tabs-4"
    className="react-tabs__tab-panel"
    id="react-tabs-5"
    role="tabpanel"
    style={Object {}}
  />
  <div
    aria-labelledby="react-tabs-6"
    className="react-tabs__tab-panel"
    id="react-tabs-7"
    role="tabpanel"
    style={Object {}}
  />
</div>
`;

exports[`<Tabs /> performance should render all tabs if forceRenderTabPanel is true 1`] = `
<div
  className="react-tabs"
  data-tabs={true}
  onClick={[Function]}
  onKeyDown={[Function]}
>
  <ul
    className="react-tabs__tab-list"
    role="tablist"
  >
    <li
      aria-controls="react-tabs-1"
      aria-disabled="false"
      aria-selected="true"
      className="react-tabs__tab react-tabs__tab--selected"
      id="react-tabs-0"
      role="tab"
      tabIndex="0"
    >
      Foo
    </li>
    <li
      aria-controls="react-tabs-3"
      aria-disabled="false"
      aria-selected="false"
      className="react-tabs__tab"
      id="react-tabs-2"
      role="tab"
      tabIndex={null}
    >
      Bar
    </li>
    <li
      aria-controls="react-tabs-5"
      aria-disabled="false"
      aria-selected="false"
      className="react-tabs__tab"
      id="react-tabs-4"
      role="tab"
      tabIndex={null}
    >
      <a>
        Baz
      </a>
    </li>
    <li
      aria-controls="react-tabs-7"
      aria-disabled="true"
      aria-selected="false"
      className="react-tabs__tab react-tabs__tab--disabled"
      id="react-tabs-6"
      role="tab"
      tabIndex={null}
    >
      Qux
    </li>
  </ul>
  <div
    aria-labelledby="react-tabs-0"
    className="react-tabs__tab-panel react-tabs__tab-panel--selected"
    id="react-tabs-1"
    role="tabpanel"
    style={Object {}}
  >
    Hello Foo
  </div>
  <div
    aria-labelledby="react-tabs-2"
    className="react-tabs__tab-panel"
    id="react-tabs-3"
    role="tabpanel"
    style={Object {}}
  >
    Hello Bar
  </div>
  <div
    aria-labelledby="react-tabs-4"
    className="react-tabs__tab-panel"
    id="react-tabs-5"
    role="tabpanel"
    style={Object {}}
  >
    Hello Baz
  </div>
  <div
    aria-labelledby="react-tabs-6"
    className="react-tabs__tab-panel"
    id="react-tabs-7"
    role="tabpanel"
    style={Object {}}
  >
    Hello Qux
  </div>
</div>
`;

exports[`<Tabs /> props should accept className 1`] = `
<div
  className="foobar"
  data-tabs={true}
  onClick={[Function]}
  onKeyDown={[Function]}
>
  <ul
    className="react-tabs__tab-list"
    role="tablist"
  >
    <li
      aria-controls="react-tabs-1"
      aria-disabled="false"
      aria-selected="true"
      className="react-tabs__tab react-tabs__tab--selected"
      id="react-tabs-0"
      role="tab"
      tabIndex="0"
    >
      Foo
    </li>
    <li
      aria-controls="react-tabs-3"
      aria-disabled="false"
      aria-selected="false"
      className="react-tabs__tab"
      id="react-tabs-2"
      role="tab"
      tabIndex={null}
    >
      Bar
    </li>
    <li
      aria-controls="react-tabs-5"
      aria-disabled="false"
      aria-selected="false"
      className="react-tabs__tab"
      id="react-tabs-4"
      role="tab"
      tabIndex={null}
    >
      <a>
        Baz
      </a>
    </li>
    <li
      aria-controls="react-tabs-7"
      aria-disabled="true"
      aria-selected="false"
      className="react-tabs__tab react-tabs__tab--disabled"
      id="react-tabs-6"
      role="tab"
      tabIndex={null}
    >
      Qux
    </li>
  </ul>
  <div
    aria-labelledby="react-tabs-0"
    className="react-tabs__tab-panel react-tabs__tab-panel--selected"
    id="react-tabs-1"
    role="tabpanel"
    style={Object {}}
  >
    Hello Foo
  </div>
  <div
    aria-labelledby="react-tabs-2"
    className="react-tabs__tab-panel"
    id="react-tabs-3"
    role="tabpanel"
    style={Object {}}
  />
  <div
    aria-labelledby="react-tabs-4"
    className="react-tabs__tab-panel"
    id="react-tabs-5"
    role="tabpanel"
    style={Object {}}
  />
  <div
    aria-labelledby="react-tabs-6"
    className="react-tabs__tab-panel"
    id="react-tabs-7"
    role="tabpanel"
    style={Object {}}
  />
</div>
`;

exports[`<Tabs /> props should have sane defaults 1`] = `
<div
  className="react-tabs"
  data-tabs={true}
  onClick={[Function]}
  onKeyDown={[Function]}
>
  <ul
    className="react-tabs__tab-list"
    role="tablist"
  >
    <li
      aria-controls="react-tabs-1"
      aria-disabled="false"
      aria-selected="true"
      className="react-tabs__tab react-tabs__tab--selected"
      id="react-tabs-0"
      role="tab"
      tabIndex="0"
    >
      Foo
    </li>
    <li
      aria-controls="react-tabs-3"
      aria-disabled="false"
      aria-selected="false"
      className="react-tabs__tab"
      id="react-tabs-2"
      role="tab"
      tabIndex={null}
    >
      Bar
    </li>
    <li
      aria-controls="react-tabs-5"
      aria-disabled="false"
      aria-selected="false"
      className="react-tabs__tab"
      id="react-tabs-4"
      role="tab"
      tabIndex={null}
    >
      <a>
        Baz
      </a>
    </li>
    <li
      aria-controls="react-tabs-7"
      aria-disabled="true"
      aria-selected="false"
      className="react-tabs__tab react-tabs__tab--disabled"
      id="react-tabs-6"
      role="tab"
      tabIndex={null}
    >
      Qux
    </li>
  </ul>
  <div
    aria-labelledby="react-tabs-0"
    className="react-tabs__tab-panel react-tabs__tab-panel--selected"
    id="react-tabs-1"
    role="tabpanel"
    style={Object {}}
  >
    Hello Foo
  </div>
  <div
    aria-labelledby="react-tabs-2"
    className="react-tabs__tab-panel"
    id="react-tabs-3"
    role="tabpanel"
    style={Object {}}
  />
  <div
    aria-labelledby="react-tabs-4"
    className="react-tabs__tab-panel"
    id="react-tabs-5"
    role="tabpanel"
    style={Object {}}
  />
  <div
    aria-labelledby="react-tabs-6"
    className="react-tabs__tab-panel"
    id="react-tabs-7"
    role="tabpanel"
    style={Object {}}
  />
</div>
`;

exports[`<Tabs /> props should honor negative defaultIndex prop 1`] = `
<div
  className="react-tabs"
  data-tabs={true}
  onClick={[Function]}
  onKeyDown={[Function]}
>
  <ul
    className="react-tabs__tab-list"
    role="tablist"
  >
    <li
      aria-controls="react-tabs-1"
      aria-disabled="false"
      aria-selected="false"
      className="react-tabs__tab"
      id="react-tabs-0"
      role="tab"
      tabIndex={null}
    >
      Foo
    </li>
    <li
      aria-controls="react-tabs-3"
      aria-disabled="false"
      aria-selected="false"
      className="react-tabs__tab"
      id="react-tabs-2"
      role="tab"
      tabIndex={null}
    >
      Bar
    </li>
    <li
      aria-controls="react-tabs-5"
      aria-disabled="false"
      aria-selected="false"
      className="react-tabs__tab"
      id="react-tabs-4"
      role="tab"
      tabIndex={null}
    >
      <a>
        Baz
      </a>
    </li>
    <li
      aria-controls="react-tabs-7"
      aria-disabled="true"
      aria-selected="false"
      className="react-tabs__tab react-tabs__tab--disabled"
      id="react-tabs-6"
      role="tab"
      tabIndex={null}
    >
      Qux
    </li>
  </ul>
  <div
    aria-labelledby="react-tabs-0"
    className="react-tabs__tab-panel"
    id="react-tabs-1"
    role="tabpanel"
    style={Object {}}
  />
  <div
    aria-labelledby="react-tabs-2"
    className="react-tabs__tab-panel"
    id="react-tabs-3"
    role="tabpanel"
    style={Object {}}
  />
  <div
    aria-labelledby="react-tabs-4"
    className="react-tabs__tab-panel"
    id="react-tabs-5"
    role="tabpanel"
    style={Object {}}
  />
  <div
    aria-labelledby="react-tabs-6"
    className="react-tabs__tab-panel"
    id="react-tabs-7"
    role="tabpanel"
    style={Object {}}
  />
</div>
`;

exports[`<Tabs /> props should honor positive defaultIndex prop 1`] = `
<div
  className="react-tabs"
  data-tabs={true}
  onClick={[Function]}
  onKeyDown={[Function]}
>
  <ul
    className="react-tabs__tab-list"
    role="tablist"
  >
    <li
      aria-controls="react-tabs-1"
      aria-disabled="false"
      aria-selected="false"
      className="react-tabs__tab"
      id="react-tabs-0"
      role="tab"
      tabIndex={null}
    >
      Foo
    </li>
    <li
      aria-controls="react-tabs-3"
      aria-disabled="false"
      aria-selected="true"
      className="react-tabs__tab react-tabs__tab--selected"
      id="react-tabs-2"
      role="tab"
      tabIndex="0"
    >
      Bar
    </li>
    <li
      aria-controls="react-tabs-5"
      aria-disabled="false"
      aria-selected="false"
      className="react-tabs__tab"
      id="react-tabs-4"
      role="tab"
      tabIndex={null}
    >
      <a>
        Baz
      </a>
    </li>
    <li
      aria-controls="react-tabs-7"
      aria-disabled="true"
      aria-selected="false"
      className="react-tabs__tab react-tabs__tab--disabled"
      id="react-tabs-6"
      role="tab"
      tabIndex={null}
    >
      Qux
    </li>
  </ul>
  <div
    aria-labelledby="react-tabs-0"
    className="react-tabs__tab-panel"
    id="react-tabs-1"
    role="tabpanel"
    style={Object {}}
  />
  <div
    aria-labelledby="react-tabs-2"
    className="react-tabs__tab-panel react-tabs__tab-panel--selected"
    id="react-tabs-3"
    role="tabpanel"
    style={Object {}}
  >
    Hello Bar
  </div>
  <div
    aria-labelledby="react-tabs-4"
    className="react-tabs__tab-panel"
    id="react-tabs-5"
    role="tabpanel"
    style={Object {}}
  />
  <div
    aria-labelledby="react-tabs-6"
    className="react-tabs__tab-panel"
    id="react-tabs-7"
    role="tabpanel"
    style={Object {}}
  />
</div>
`;

exports[`<Tabs /> should not add known props to dom 1`] = `
<div
  className="react-tabs"
  data-tabs={true}
  onClick={[Function]}
  onKeyDown={[Function]}
/>
`;

exports[`<Tabs /> should pass through custom properties 1`] = `
<div
  className="react-tabs"
  data-tabs={true}
  data-tooltip="Tooltip contents"
  onClick={[Function]}
  onKeyDown={[Function]}
/>
`;

exports[`<Tabs /> validation should allow other DOM nodes 1`] = `
<div
  className="react-tabs"
  data-tabs={true}
  onClick={[Function]}
  onKeyDown={[Function]}
>
  <div
    id="tabs-nav-wrapper"
  >
    <button>
      Left
    </button>
    <div
      className="tabs-container"
    >
      <ul
        className="react-tabs__tab-list"
        role="tablist"
      >
        <li
          aria-controls="react-tabs-1"
          aria-disabled="false"
          aria-selected="true"
          className="react-tabs__tab react-tabs__tab--selected"
          id="react-tabs-0"
          role="tab"
          tabIndex="0"
        />
        <li
          aria-controls="react-tabs-3"
          aria-disabled="false"
          aria-selected="false"
          className="react-tabs__tab"
          id="react-tabs-2"
          role="tab"
          tabIndex={null}
        />
      </ul>
    </div>
    <button>
      Right
    </button>
  </div>
  <div
    className="tab-panels"
  >
    <div
      aria-labelledby="react-tabs-0"
      className="react-tabs__tab-panel react-tabs__tab-panel--selected"
      id="react-tabs-1"
      role="tabpanel"
      style={Object {}}
    />
    <div
      aria-labelledby="react-tabs-2"
      className="react-tabs__tab-panel"
      id="react-tabs-3"
      role="tabpanel"
      style={Object {}}
    />
  </div>
</div>
`;

exports[`<Tabs /> validation should allow random order for elements 1`] = `
<div
  className="react-tabs"
  data-tabs={true}
  onClick={[Function]}
  onKeyDown={[Function]}
>
  <div
    aria-labelledby="react-tabs-0"
    className="react-tabs__tab-panel react-tabs__tab-panel--selected"
    id="react-tabs-1"
    role="tabpanel"
    style={Object {}}
  >
    Hello Foo
  </div>
  <ul
    className="react-tabs__tab-list"
    role="tablist"
  >
    <li
      aria-controls="react-tabs-1"
      aria-disabled="false"
      aria-selected="true"
      className="react-tabs__tab react-tabs__tab--selected"
      id="react-tabs-0"
      role="tab"
      tabIndex="0"
    >
      Foo
    </li>
    <li
      aria-controls="react-tabs-3"
      aria-disabled="false"
      aria-selected="false"
      className="react-tabs__tab"
      id="react-tabs-2"
      role="tab"
      tabIndex={null}
    >
      Bar
    </li>
  </ul>
  <div
    aria-labelledby="react-tabs-2"
    className="react-tabs__tab-panel"
    id="react-tabs-3"
    role="tabpanel"
    style={Object {}}
  >
    Hello Bar
  </div>
</div>
`;
